src/coreclr/nativeaot/Runtime/unix/unixasmmacrosarm64.inc

```diff
@@ -24,9 +31,11 @@ C_FUNC(\Name):
 .endm
 
 .macro ALTERNATE_ENTRY Name
-        .global C_FUNC(\Name)
 #if !defined(__APPLE__)
+        .global C_FUNC(\Name)
         .hidden C_FUNC(\Name)
+#else
+        .alt_entry C_FUNC(\Name)
 #endif
```

Comment:
Should it be also done for NO_CFI_EPILOG?

```c
#ifndef NO_CFI_EPILOG
        .global C_FUNC(\Name)
#endif
#if !defined(__APPLE__)
        .hidden C_FUNC(\Name)
#elif defined(NO_CFI_EPILOG)
        .alt_entry C_FUNC(\Name)
#endif
```

Reply:
This should be done unconditionally. There are certain broken Clang versions where `.alt_entry` triggers errors but AFAIK it should not affect last couple of betas or any released version.

Reply:
ok. In theory this should fix the CI build but I haven't tested.

Reply:
Unlike ELF which can support pretty much arbitrary number of sections, the Mach-O object files are limited to ~255 sections. Apple uses the `subsections-via-symbols` linking approach to circumvent this limitation. In the `subsections-via-symbols` model the global symbols are used to slice the section into subsections (called atoms) which are then used as the linking unit (for purpose of dead stripping, possible reordering, etc.). Historically you had to opt-in into it but any Clang compiler in the past decade did so. The new linker included in Xcode 15 basically broke the support for object files without `subsections-via-symbols`, and the old linker is now scheduled to be completely removed.

`.alt_entry` is an escape hatch to introduce symbols that are globally visible, but at the same time don't cause the section/method to be split into two separate atoms. It's the only supported way to place globally visible labels into middle of the method, which is used by NativeAOT to mark instructions that (legally) cause SIGSEGVs.

In Apple Clang 16 (LLVM 17? 18?) there were additional checks added to prevent the misuse of global labels. Likewise, checks were added for misuse of the CFI opcodes. We need to correct the prior mistake of relying on undefined behavior in the linker that kept the assembly code order intact. That's why `.alt_entry` is unconditional. Unfortunately, there was a code in Clang that didn't correctly interpret `.alt_entry` for the purpose of consistency checks of the atoms and the CFI unwind information. It resulted in `.alt_entry` being incorrectly interpreted as a global label that splits atoms and any negative (relative) offsets in CFI epilogs trigger an error because it thinks the whole method goes into negative (absolute) offsets in CFI.